### PR TITLE
CUDA: support for weight clamp in top-k norm

### PR DESCRIPTION
**Pull request by am17an**
Support #16655

### CODE REVIEW

**Comment by am17an** [comment]
@CISC maybe need to rebase? Idk what happened

**Comment by ORippler** [comment]
> @CISC maybe need to rebase? Idk what happened

Stacked PRs + squash-merge require linear history to work well. So the "best" way to do this is to no longer do [merges](https://github.com/ggml-org/llama.cpp/pull/16655/commits/73208422f5afa7588809e29d023d6ff6f93cdae0) to include upstream changes into a branch, but only do sequential, (interactive) rebases + force pushes from the bottom to the top of the stacked PR chain.

If you:
* want and need [6de8ed7](https://github.com/ggml-org/llama.cpp/pull/16702/commits/6de8ed75196c7cd98c1f34bbf3a7452451ba8ac2)..[03792ad](https://github.com/ggml-org/llama.cpp/pull/16702/commits/03792ad93609fc67e41041c6347d9aa14e5e0d74) in https://github.com/ggml-org/llama.cpp/tree/cisc/norm-topk-bias , you need to rebase https://github.com/ggml-org/llama.cpp/tree/cisc/norm-topk-bias onto master and then do an interactive rebase of am17an:norm-topk-cuda onto https://github.com/ggml-org/llama.cpp/tree/cisc/norm-topk-bias, keeping only [4267d3e](https://github.com/ggml-org/llama.cpp/pull/16702/commits/4267d3e9de0213cd6f09f0bb96ef5ec2ef05e94f)..[161248c](https://github.com/ggml-org/llama.cpp/pull/16702/commits/161248c5f3ecf05077b4c199ee24687be6777fec) (i.e. discarding the range you need in your target branch)
* do an interactive rebase onto https://github.com/ggml-org/llama.cpp/tree/cisc/norm-topk-bias, discarding the commit range mentioned above if you do not want and need them. This is likely to be the case if you spun your branch off of master instead of https://github.com/ggml-org/llama.cpp/tree/cisc/norm-topk-bias

Once again: Doing stacked PRs manually with git is possible but cumbersome, which is why there exists a plethora of tooling to [manage stacked PRs in git](https://www.stacking.dev/)

**Comment by CISC** [comment]
It looks like this happened because you pulled in master, hopefully it is recoverable though, but if not just start a new one.

**Comment by am17an** [comment]
> It looks like this happened because you pulled in master, hopefully it is recoverable though, but if not just start a new one.

It is recoverable, but your branch needs to be updated. For now I can create this PR on master including your change, after you merge your PR we can merge this one

**Review by jeffbolznv — COMMENTED** [review]
You need to either check that the upper bound of the clamp is INFINITY, or just always do both sides of the clamp.

File: ggml/src/ggml-cuda/topk-moe.cu [hunk 1]
```diff
@@ -242,13 +250,21 @@ void ggml_cuda_op_topk_moe(ggml_backend_cuda_context & ctx,
 
     const int n_expert_used = weights->ne[1];
 
+    float clamp_val = 0.0f;
     if (with_norm) {
-        launch_topk_moe_cuda<true>(ctx, logits_d, weights_d, ids_d, n_rows, n_experts, n_expert_used);
+        if (clamp) {
+            clamp_val = ggml_get_op_params_f32(clamp, 0);
+            float max_val = ggml_get_op_params_f32(clamp, 1);
+            GGML_ASSERT(max_val == INFINITY);
```
Comment (jeffbolznv) [inline ggml/src/ggml-cuda/topk-moe.cu]: An assert isn't sufficient. The backend needs to correctly handle whatever input it's given.

Comment (am17an) [reply ggml/src/ggml-cuda/topk-moe.cu]: This can't be changed via the user anyway. And tests will not pass if someone changes the value in llama-graph where it is hardcoded. As such this clamp is for avoiding div by zero, adding the upper side clamp is just a wasted instruction IMO.

Comment (jeffbolznv) [reply ggml/src/ggml-cuda/topk-moe.cu]: ggml is a general-purpose library and people can generate whatever graph they like. The backend needs to handle any valid input, not just what llama.cpp generates.

You should think of fusion like compiler optimizations - it needs to check that the input exactly matches its expectations, and either handle the input or skip the optimization.

Comment (am17an) [reply ggml/src/ggml-cuda/topk-moe.cu]: Okay fair point. I'll make the change for the upper bound to be inf or skip the fusion

File: ggml/src/ggml-cuda/topk-moe.cu [hunk 2]
```diff
@@ -242,18 +250,24 @@ void ggml_cuda_op_topk_moe(ggml_backend_cuda_context & ctx,
 
     const int n_expert_used = weights->ne[1];
 
+    float clamp_val = 0.0f;
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/topk-moe.cu]: Isn't this default value problematic? The clamping in the kernel is being done unconditionally so won't this result in all positive values being clamped to 0 if `clamp == nullptr`?

File: ggml/src/ggml-cuda/topk-moe.cu [hunk 3]
```diff
@@ -242,18 +250,24 @@ void ggml_cuda_op_topk_moe(ggml_backend_cuda_context & ctx,
...
     if (with_norm) {
-        launch_topk_moe_cuda<true>(ctx, logits_d, weights_d, ids_d, n_rows, n_experts, n_expert_used);
+        if (clamp) {
+            clamp_val = ggml_get_op_params_f32(clamp, 0);
+        }
+        launch_topk_moe_cuda<true>(ctx, logits_d, weights_d, ids_d, n_rows, n_experts, n_expert_used, clamp_val);
     } else {
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/topk-moe.cu:259]: The else branch needs an assert that `clamp == nullptr` since the kernel lacks support.

File: ggml/src/ggml-cuda/topk-moe.cu (hunk 2 above)
Comment (am17an) [reply ggml/src/ggml-cuda/topk-moe.cu]: The clamp is on the lower bound, if clamp == nullptr, then it does `val = max(0, val)`, perhaps not named appropriately

Comment (jeffbolznv) [reply ggml/src/ggml-cuda/topk-moe.cu]: I think Johannes is suggesting you use -INFINITY as a default. Maybe it wouldn't actually matter in practice if softmax makes everything positive, but it would be less confusing.

Comment (JohannesGaessler) [reply ggml/src/ggml-cuda/topk-moe.cu]: Just add a comment explaining that the values are bounded to the interval (0, 1) and I think it's fine.

**Review by JohannesGaessler — APPROVED** [review]
